ocaml/xapi/xapi_pci.ml

```diff
@@ -319,3 +319,14 @@ let get_system_display_device () =
       )
       None items
   with _ -> None
+
+let disable_dom0_access ~__context ~self =
+  ignore
+  @@ Xapi_pci_helpers.update_dom0_access ~__context ~pci:self ~action:`disable
+
+let enable_dom0_access ~__context ~self =
+  ignore
```

Comment:
Can these functions return the new value, like the pgpu function does?

Reply:
Yes I can do that, makes more sense indeed.

Reply:
Hmm actually that would make me move `pgpu_dom0_access` enum in the PCI module, i'm not sure it makes sens since the DB field is at pgpu level :thinking:

Reply:
The enum can be defined at line 6518, right after the comment "PCI Devices", then the defnition can be shared by both sets of functions. The enum can also be renamed to `pci_dom0_access` to make sure it's related to pci devices in general.

```patch
diff --git a/ocaml/idl/datamodel.ml b/ocaml/idl/datamodel.ml
index aaf4157e0..15a34eccf 100644
--- a/ocaml/idl/datamodel.ml
+++ b/ocaml/idl/datamodel.ml
@@ -6518,6 +6518,21 @@ end
 
 (** PCI devices *)
 
+let pci_dom0_access =
+  Enum
+    ( "pgpu_dom0_access"
+    , [
+        ("enabled", "dom0 can access this device as normal")
+      ; ( "disable_on_reboot"
+        , "On host reboot dom0 will be blocked from accessing this device"
+        )
+      ; ("disabled", "dom0 cannot access this device")
+      ; ( "enable_on_reboot"
+        , "On host reboot dom0 will be allowed to access this device"
+        )
+      ]
+    )
+
 module PCI = struct
   let disable_dom0_access =
     call ~name:"disable_dom0_access" ~lifecycle:[]
@@ -6525,6 +6540,7 @@ module PCI = struct
         "Hide a PCI device from the dom0 kernel. (Takes affect after next \
          boot.)"
       ~params:[(Ref _pci, "self", "The PCI to hide")]
+      ~result:(pci_dom0_access, "The accessibility of this PCI device from dom0")
       ~allowed_roles:_R_POOL_OP ()
 
   let enable_dom0_access =
@@ -6533,6 +6549,7 @@ module PCI = struct
         "Unhide a PCI device from the dom0 kernel. (Takes affect after next \
          boot.)"
       ~params:[(Ref _pci, "self", "The PCI to unhide")]
+      ~result:(pci_dom0_access, "The accessibility of this PCI device from dom0")
       ~allowed_roles:_R_POOL_OP ()
 
   let is_dom0_access_enabled =
@@ -6645,21 +6662,6 @@ end
 (** Physical GPUs (pGPU) *)
 
 module PGPU = struct
-  let dom0_access =
-    Enum
-      ( "pgpu_dom0_access"
-      , [
-          ("enabled", "dom0 can access this device as normal")
-        ; ( "disable_on_reboot"
-          , "On host reboot dom0 will be blocked from accessing this device"
-          )
-        ; ("disabled", "dom0 cannot access this device")
-        ; ( "enable_on_reboot"
-          , "On host reboot dom0 will be allowed to access this device"
-          )
-        ]
-      )
-
   let add_enabled_VGPU_types =
     call ~name:"add_enabled_VGPU_types"
       ~lifecycle:[(Published, rel_vgpu_tech_preview, "")]
@@ -6791,7 +6793,7 @@ module PGPU = struct
           ; param_default= None
           }
         ]
-      ~result:(dom0_access, "The accessibility of this PGPU from dom0")
+      ~result:(pci_dom0_access, "The accessibility of this PGPU from dom0")
       ~allowed_roles:_R_POOL_OP ()
 
   let disable_dom0_access =
@@ -6807,7 +6809,7 @@ module PGPU = struct
           ; param_default= None
           }
         ]
-      ~result:(dom0_access, "The accessibility of this PGPU from dom0")
+      ~result:(pci_dom0_access, "The accessibility of this PGPU from dom0")
       ~allowed_roles:_R_POOL_OP ()
 
   let t =
@@ -6868,7 +6870,7 @@ module PGPU = struct
             "A map relating each VGPU type supported on this GPU to the \
              maximum number of VGPUs of that type which can run simultaneously \
              on this GPU"
-        ; field ~qualifier:DynamicRO ~ty:dom0_access
+        ; field ~qualifier:DynamicRO ~ty:pci_dom0_access
             ~lifecycle:[(Published, rel_cream, "")]
             ~default_value:(Some (VEnum "enabled")) "dom0_access"
             "The accessibility of this device from dom0"
diff --git a/ocaml/xapi/xapi_pci.ml b/ocaml/xapi/xapi_pci.ml
index 14931b81f..326babcfa 100644
--- a/ocaml/xapi/xapi_pci.ml
+++ b/ocaml/xapi/xapi_pci.ml
@@ -321,12 +321,10 @@ let get_system_display_device () =
   with _ -> None
 
 let disable_dom0_access ~__context ~self =
-  ignore
-  @@ Xapi_pci_helpers.update_dom0_access ~__context ~pci:self ~action:`disable
+  Xapi_pci_helpers.update_dom0_access ~__context ~pci:self ~action:`disable
 
 let enable_dom0_access ~__context ~self =
-  ignore
-  @@ Xapi_pci_helpers.update_dom0_access ~__context ~pci:self ~action:`enable
+  Xapi_pci_helpers.update_dom0_access ~__context ~pci:self ~action:`enable
 
 let is_dom0_access_enabled ~__context ~self =
   not (Pciops.is_pci_hidden ~__context self)
diff --git a/ocaml/xapi/xapi_pci.mli b/ocaml/xapi/xapi_pci.mli
index af96037ba..09816099e 100644
--- a/ocaml/xapi/xapi_pci.mli
+++ b/ocaml/xapi/xapi_pci.mli
@@ -52,11 +52,17 @@ val disable_system_display_device : unit -> unit
 val dequarantine : __context:Context.t -> Xenops_interface.Pci.address -> unit
 (** dequarantine a PCI device. This is idempotent. *)
 
-val disable_dom0_access : __context:Context.t -> self:API.ref_PCI -> unit
-(** Hide a PCI device from the dom0 kernel. (Takes affect after next boot.) *)
+val disable_dom0_access :
+     __context:Context.t
+  -> self:API.ref_PCI
+  -> [`disable_on_reboot | `disabled | `enable_on_reboot | `enabled]
+(** Hide a PCI device from the dom0 kernel *)
 
-val enable_dom0_access : __context:Context.t -> self:API.ref_PCI -> unit
-(** Unhide a PCI device from the dom0 kernel. (Takes affect after next boot.) *)
+val enable_dom0_access :
+     __context:Context.t
+  -> self:API.ref_PCI
+  -> [`disable_on_reboot | `disabled | `enable_on_reboot | `enabled]
+(** Unhide a PCI device from the dom0 kernel *)
 
 val is_dom0_access_enabled : __context:Context.t -> self:API.ref_PCI -> bool
 (** Check whether a PCI device will be hidden from the dom0 kernel on boot. *)
```

Reply:
Fine by me :)